portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -49,30 +50,42 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     int requestLimit = Integer.parseInt(request.getParameter("request-limit"));
     String language = request.getParameter("language");
 
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
-
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    PreparedQuery results = datastore.prepare(query);
-
-    List<Comment> comments = new ArrayList<>();
-    for (Entity entity : results.asIterable(FetchOptions.Builder.withLimit(requestLimit))) {
-      String author = (String) entity.getProperty("author");
-      String text = (String) entity.getProperty("text");
-
-      Translate translate = TranslateOptions.newBuilder().setProjectId("internplayground")
-          .setQuotaProjectId("internplayground").build().getService();
-      Translation translation = translate.translate(text,
-          Translate.TranslateOption.targetLanguage(language));
-      text = translation.getTranslatedText();
-      Comment comment = new Comment(author, text);
-      comments.add(comment);
+    Translate translate = TranslateOptions.newBuilder().setProjectId("internplayground")
+        .setQuotaProjectId("internplayground").build().getService();
+
+    List<Language> languages = translate.listSupportedLanguages();
```

Comment:
Could you use:
Boolean langParamValid = languages.contains(lang.getCode())

here?

Reply:
How does that work? Isn't `lang` only defined inside the for loop?

Reply:
Yes. The joys of too many similarly named variables; the point of the question is whether you can use the language request parameter in a way that doesn't involve a linear search.

The specific question was intentionally flawed differently, but was meant to make room to think:

Boolean langParamValid = languages.contains(language)

Won't work, because List<Language> doesn't contain a String.

What about:
Boolean langParamValid languages.stream().filter(l -> l.getCode().equals(language)).findFirst().isPresent();

Using the Java 8 stream/filter for a List. Not a loop, but done like this, probably slower... We can add a .parallel() in there to speed it up though...

But, we can do better!

Why should we query the API for the supported languages on every page reload and post? Once the ProvideLanguageServlert is up and running, can we use some caching to call it once, keep a result, and perhaps even keep it in a such a way as to allow an efficient lookup here?

Potentially, resulting in something where the answer to the original question (sans incorrect variable usage) might work,  or ...:

Boolean langParamValid = supportedLanguages.containsKey(language)

Soooo many options!


The real point of this question was to attempt to draw your thinking to time complexity:
https://en.wikipedia.org/wiki/Time_complexity

A for loop represents a linear search, O(n). Using the Java 8 streams would be a parallel search, which would still be linear, but perhaps O(n/2). A HashMap lookup is often considered to be O(1), ie, no matter how large the list is, it only needs a single operation to find a value.

I'd suggest keeping this as a for loop for now, the data set is well bounded, and small. However, keep in mind that caching results, and choice of data structures, can have significant impact on performance (and reliability) as a design needs to start scaling.

Reply:
Yeah, I had time complexity in mind when I wrote it, but I couldn't think of a more efficient way beyond breaking out of the loop early.